Generate `run` stm function from gospel information

n-osborne opened the pull request:
This PR is a one commit PR based on #12 (ignore all but the last commit).

As for `postcond`, exception raising is not supported yet.
The generation of `run` is not monadic, as all the potential errors should have been caught in an earlier stage.

n-osborne committed:
Generate `run` stm functions.

As for `postconditions`, exception raising is not supported yet.

shym commented on plugins/qcheck-stm/src/stm_of_ir.ml:
```diff
@@ -120,6 +120,65 @@ let mk_cmd_pattern value =
   let name = String.capitalize_ascii (str_of_ident value.id) |> lident in
   ppat_construct name args
 
+let ty_show_of_core_type inst ty =
```
This function could be structured (and named) closer to the equivalent one for `next_state`, `mk_pat_ret_ty`. In particular it would be nice to put them close together and to reuse `munge_longident` (pulling the `capitalize_ascii` out as an argument, I suppose), as they should do the same thing apart from the pat/exp target.

shym commented on plugins/qcheck-stm/src/stm_of_ir.ml:
```diff
@@ -120,6 +120,65 @@ let mk_cmd_pattern value =
+
+let exp_of_ident id = pexp_ident (lident (str_of_ident id))
+
+let call config esut value =
```
Will that function be used elsewhere than in `run_case`? I’d propose to move it inside `run_case` and/or give it a name that’s easier to decode without context (`function_call_exp`?).

shym reviewed (commented):
I suggest two small changes to increase the uniformity (and help maintenance :-).

shym reviewed (commented):
I suggest two small changes to increase the uniformity (and help maintenance :-).

n-osborne committed:
Add `Reserr.app`

A monad is also an applicative functor.

n-osborne committed:
Refactorize transformations of core_type

- `mk_pat_of_ret_ty` becomes `pat_of_core_type`
- `ty_show_of_core_type` becomes `exp_of_core_type`
- both functions Return_have the same structure
- generalize `munge_longident`

Co-authored-by: Samuel Hym <samuel.hym@rustyne.lautre.net>

shym reviewed (approved):
This is ready to merge now!
